pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,200 @@
+
+    last_updated_by = attr.ib(default=None, type=str)
+    """Person/party who updated the report last time."""
+
+    entries = attr.ib(default=attr.Factory(tuple), type=tuple)
+    """A tuple of :class:`MaintenanceEntry` objects, indicating
+    which repositories are in maintenance mode and details.
+    If empty, then it means no repositories are in maintenance mode.
+    """
+
+    @entries.validator
+    def check_duplicates(self, attribute, value):
+        # pylint: disable=unused-argument
+        # check if there's duplicate entries
+        repo_ids = [entry.repo_id for entry in value]
+        if len(repo_ids) != len(set(repo_ids)):
+            raise ValueError("Duplicate entries")
+
+    @classmethod
+    def _from_data(cls, data):
+        """Create a new report with raw data
+
+        Args:
+            data (dict):
+                A dict containing a raw representation of the maintenance status.
+        Returns:
+            a new instance of ``cls``
+
+        Raises:
+            InvalidDataException
+                If the provided ``data`` fails validation against an expected schema.
+        """
+        try:
+            jsonschema.validate(instance=data, schema=cls._SCHEMA)
+        except jsonschema.exceptions.ValidationError as error:
+            LOG.exception("%s.from_data invoked with invalid data", cls.__name__)
+            raise InvalidDataException(str(error))
+
+        entries = []
+        for repo_id, details in data["repos"].items():
+            entries.append(MaintenanceEntry(repo_id=repo_id, **details))
+
+        maintenance = cls(
+            last_updated=data["last_updated"],
+            last_updated_by=data["last_updated_by"],
+            entries=tuple(entries),
+        )
+
+        return maintenance
+
+    def _export_dict(self):
+        """export a raw dictionary of maintenance report"""
+        report = {
+            "last_updated": self.last_updated,
+            "last_updated_by": self.last_updated_by,
+            "repos": {},
+        }
+
+        for entry in self.entries:
+            report["repos"].update(
+                {
+                    entry.repo_id: {
+                        "message": entry.message,
+                        "owner": entry.owner,
+                        "started": entry.started,
+                    }
+                }
+            )
+
+        return report
+
+    def add(self, repo_ids, **kwargs):
+        """Add entries to maintenance report and update the timestamp. Every
+        entry added to the report represents a repository in maintenace mode.
+
+        Args:
+            repo_ids (list[str]):
+                A list of repository ids. New entries with these repository ids will
+                be added to the maintenance report.
+
+                Note: it's users' responsibility to make sure the repository exists in
+                the pulp server, this method doesn't check the repository's existence.
```

Comment:
"repository's existence" is technically wrong grammar for referring to multiple repositories.  It's a bit awkward to get the correct wording for this in English.  Also, Pulp should use capital P.  Here's my attempt:

```suggestion
                the Pulp server, this method doesn't check for the existence of repositories.
```